ReactiveSwift.playground/Pages/Sandbox.xcplaygroundpage/Contents.swift

```diff
@@ -22,4 +22,186 @@ import Foundation
  A place where you can build your sand castles 🏖.
 */
 
+enum PromiseResult<Value, Error: Swift.Error> {
+    case fulfilled(Value)
+    case rejected(Error?)   // nil error means that the producer was interrupted or empty
+    
+    init(value: Value) {
+        self = .fulfilled(value)
+    }
+    
+    init(error: Error?) {
+        self = .rejected(error)
+    }
+    
+    var value: Value? {
+        switch self {
+        case .fulfilled(let value): return value
+        case .rejected(_): return nil
+        }
+    }
+    
+    var error: Error? {
+        switch self {
+        case .fulfilled(_): return nil
+        case .rejected(let error): return error
+        }
+    }
+}
+
+protocol Thenable {
```

Comment:
I'm not sure we'd get any benefit from this protocol.

Reply:
Yeah.. I just snatched it from PromiseKit, after [reading the specs](https://promisesaplus.com/) which refer to  `Thenable` as something separate from the `Promise` itself. Technically, there could be thin auxiliary classes implementing the protocol directly, without the overhead of the full `Promise`

Reply:
I have refactored the code, and now Thenable is a type-erased ThenableType, which allows for a number of nice optimisations